### PR TITLE
Add support for cert-manager v1.3 Approval conditions

### PR DESCRIPTION
**Pull request by jakexks**

### CODE REVIEW

File: config/rbac/cert_manager_controller_approver_clusterrole.yaml [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+  name: cert-manager-controller-approve:casissuer
+rules:
+- apiGroups:
+  - cert-manager.io
+  resources:
+  - signers
+  verbs:
+  - approve
+  resourceNames:
+  - cas-issuer.jetstack.io/*
```
Comment (JoshVanL) [inline config/rbac/cert_manager_controller_approver_clusterrole.yaml:13]: Needs to be:

```yaml
  - googlecasclusterissuers.cas-issuer.jetstack.io/*
  - googlecasissuers.cas-issuer.jetstack.io/*
 ```

File: pkg/controller/certificaterequest/certificaterequest_controller.go [hunk 2]
```diff
@@ -189,6 +191,26 @@ func (r *CertificateRequestReconciler) Reconcile(ctx context.Context, req ctrl.R
...
 	}
 
+	// From cert-manager v1.3 onwards, CertificateRequests must be approved before they are signed.
+	if !viper.GetBool("disable-approval-check") {
+		log.Info("Checking whether CR has been approved", "cr", req.NamespacedName)
+		// Explicitly fail if the certificate request has been denied
+		if cmutil.CertificateRequestIsDenied(&certificateRequest) {
+			msg := "certificate request has been denied, not signing"
+			log.Info(msg, "cr", req.NamespacedName)
+			setReadyCondition(cmmeta.ConditionFalse, cmapi.CertificateRequestReasonFailed, msg)
```
Comment (JoshVanL) [inline pkg/controller/certificaterequest/certificaterequest_controller.go:201]: It's been mostly decided that we are going to use a different Ready Reason than Failed https://github.com/jetstack/cert-manager/pull/3878. You might want to wait for v1.3.1 for `cmapi.CertificateRequestReasonDenied`, but you can use anything you want.

Comment (JoshVanL) [reply pkg/controller/certificaterequest/certificaterequest_controller.go:201]: We should also move this line out of the `disable-approval-check` flag check. We optimistically always check for Denial, but can choose to ignore approval.

**Review by JoshVanL — CHANGES_REQUESTED** [review]
Couple things from me.

/assign @jakexks

File: pkg/controller/certificaterequest/certificaterequest_controller.go [hunk 3]
```diff
@@ -111,6 +128,26 @@ func (r *CertificateRequestReconciler) Reconcile(ctx context.Context, req ctrl.R
 		}
 	}()
 
+	// Explicitly fail if the certificate request has been denied
+	if cmutil.CertificateRequestIsDenied(&certificateRequest) {
+		msg := "certificate request has been denied, not signing"
+		log.Info(msg, "cr", req.NamespacedName)
+		setReadyCondition(cmmeta.ConditionFalse, cmapi.CertificateRequestReasonDenied, msg)
+		r.Recorder.Event(&certificateRequest, eventTypeWarning, cmapi.CertificateRequestReasonDenied, msg)
```
Comment (JoshVanL) [inline pkg/controller/certificaterequest/certificaterequest_controller.go:136]: I think we shouldn't fire an event if the CR is denied, as the approver should fire an event anyway. Saves a bit of pressure on the API.

**Review by JoshVanL — COMMENTED** [review]
Final thing from me. You may want to hold cancel if you disagree.

/hold 
/lgtm

**Comment by jakexks** [comment]
(leaving on hold while I add tests on another branch)

**Pull request merged**